db-queue-spring/src/main/java/ru/yoomoney/tech/dbqueue/spring/dao/H2QueuePickTaskDao.java

```diff
@@ -0,0 +1,288 @@
+                (ResultSet rs) -> {
+                    if (!(rs.next() && rs.getMetaData().getColumnCount() > 1))
+                        return null;
+
+                    Map<String, String> additionalData = queueTableSchema
+                            .getExtFields()
+                            .stream()
+                            .collect(
+                                    LinkedHashMap::new,
+                                    (map, key) -> {
+                                        try {
+                                            map.put(key, rs.getString(key));
+                                        } catch (SQLException e) {
+                                            throw new RuntimeException(e);
+                                        }
+                                    }, Map::putAll);
+
+                    return TaskRecord.builder()
+                            .withId(rs.getLong(queueTableSchema.getIdField()))
+                            .withCreatedAt(getZonedDateTime(rs, queueTableSchema.getCreatedAtField()))
+                            .withNextProcessAt(getZonedDateTime(rs, queueTableSchema.getNextProcessAtField()))
+                            .withPayload(rs.getString(queueTableSchema.getPayloadField()))
+                            .withAttemptsCount(rs.getLong(queueTableSchema.getAttemptField()))
+                            .withReenqueueAttemptsCount(rs.getLong(queueTableSchema.getReenqueueAttemptField()))
+                            .withTotalAttemptsCount(rs.getLong(queueTableSchema.getTotalAttemptField()))
+                            .withExtData(additionalData).build();
+
+                }, location.getQueueId().asString(), pickTaskSettings.getRetryInterval().getSeconds());
+    }
+
+    public static class PickerProcedure {
+        private static final Map<String, PickQuery> pickTaskSqlCache = new ConcurrentHashMap<>();
+        private static final RowIdLocker rowIdLocker = new RowIdLocker();
+
+        public static ResultSet pickTask(final Connection connection,
+                                         final String queueName,
+                                         final long retryInterval) throws SQLException {
+            final PickQuery pickQuery = pickTaskSqlCache.get(queueName);
+            Objects.requireNonNull(pickQuery, "A pick query can't be null");
+
+            final Savepoint savepoint = connection.setSavepoint("start-picking-task-point");
+
+            final Long taskId = rowIdLocker.lock(queueName, rowIds -> {
+                try {
+
+                    final String sql;
+                    if (rowIds.isEmpty())
+                        sql = String.format(pickQuery.selectSql, " (1 = 1) ");
+                    else
+                        sql = String.format(
+                                pickQuery.selectSql,
+                                " _ROWID_ IN " + rowIds
```

Comment:
I cannot figure out how this should work. 
Lets pretend there are 5 threads and many tasks await processing. 
1) select ... where 1=1 for update
2) select ... where _ROWID_ in (1) for update
3) select ... where _ROWID_ in (1, 2) for update
4) select ... where _ROWID_ in (1, 2, 3) for update
5) select ... where _ROWID_ in (1, 2, 3, 4) for update

This doesn't make sense to me, because we should ignore locked rows instead of picking them up.

Reply:
My bad. It's absolutely wrong. 🤦‍♂️